vm/actor/src/builtin/init/state.rs

```diff
@@ -34,21 +36,38 @@ impl State {
         let id = self.next_id;
         self.next_id += 1;
 
-        let mut map: Hamt<String, _> = Hamt::load_with_bit_width(&self.address_map, store, 5)?;
-        map.set(String::from_utf8_lossy(&addr.to_bytes()).to_string(), id)?;
+        let mut map: Hamt<String, _> =
+            Hamt::load_with_bit_width(&self.address_map, store, HAMT_BIT_WIDTH)?;
+        map.set(addr.hash_key(), id)?;
         self.address_map = map.flush()?;
 
         Ok(Address::new_id(id.0).expect("Id Address should be created without Error"))
     }
 
-    /// Resolve address
+    /// ResolveAddress resolves an address to an ID-address, if possible.
+    /// If the provided address is an ID address, it is returned as-is.
+    /// This means that ID-addresses (which should only appear as values, not keys)
+    /// and singleton actor addresses pass through unchanged.
+    ///
+    /// Post-condition: all addresses succesfully returned by this method satisfy `addr.protocol() == Protocol::ID`.
     pub fn resolve_address<BS: BlockStore>(
         &self,
-        _store: &BS,
-        _addr: &Address,
-    ) -> Result<Address, String> {
-        // TODO implement address resolution
-        todo!()
+        store: &BS,
+        addr: &Address,
+    ) -> Result<Address, HamtError> {
+        if addr.protocol() == Protocol::ID {
+            return Ok(addr.clone());
+        }
+
+        let map: Hamt<String, _> =
+            Hamt::load_with_bit_width(&self.address_map, store, HAMT_BIT_WIDTH)?;
+
+        let actor_id: ActorID = match map.get(&addr.hash_key())? {
+            Some(id) => id,
+            None => return Err(HamtError::Custom("address not found")),
```

Comment:
Small nit: are we capitalizing error messages? Whichever it is I think we should keep consistent.